### ISSUE TITLE
Dart mojo bindings are somewhat convoluted

### ISSUE THREAD

**Issue by Hixie** [domokit/mojo#647]
It's quite hard to follow what's going on with the dart bindings. There's a lot of very similarly-named classes with very similar sets of methods.

For example, take the code generated for the Shell interface:

There's Shell, the abstract interface corresponding to the mojom interface.
_ShellProxyCalls is the implementation of the pure Shell interface.
It has a pointer to a _ShellProxyImpl, which extends bindings.Proxy.
There's ShellProxy, which implements bindings.ProxyBase, and which has a pointer to _ShellProxyImpl.

This is very confusing. The class that says it's a "Proxy" is not the class that extends "Proxy"; "ProxyBase" is not a base for "Proxy", and the implementation is called "Calls" and there's an "Impl" class that isn't the implementation. It's not clear to me why there has to be both the Proxy and the ProxyBase, either.

cc @zra, @abarth.

**Comment by zanderso** [comment]
Happy to entertain suggestions for better names if you have any.

**Comment by Hixie** [comment]
Can you explain the distinction between the Proxy subclasses and the ProxyBase subclasses?

**Comment by johnmccutchan** [comment]
I largely agree and had a bunch of similar concerns when I first started looking at the the Dart bindings. Another first impression:

- When dealing with a Proxy you must access a field called "ptr" which I guess is short hand for "pointer", which has no meaning in Dart. 

As for naming suggestions, here are the current names and what they appear to do (using Shell as an example):

```dart
abstract class bindings.Proxy extends core.MojoEventHandler {
  // An event handler that reads and writes a mojo message pipe.
  // Used to send messages with sendMessage and sendMessageWithRequestId.
  // Reads responses and calls the abstract handleResponse
}
```

```dart
abstract class bindings.Stub extends core.MojoEventHandler {
  // An event handler that reads and writes a mojo message pipe.
  // Used to send build/send responses with buildResponse, buildResponseWithId, and _sendResponse.
  // Reads messages and calls the abstract handleMessage.
}
```

```dart
abstract class bindings.ProxyBase {
  // just fields: service name, proxy impl, and ptr.
}
```

```dart
abstract class Shell {
  // The abstract Shell interface.
}
```

```dart
class _ShellProxyImpl extends bindings.Proxy {
  // Implements bindings.Proxy.handleResponse for this specific interface.
  // Responses are deserialized and are used to complete a Future.
}
```

```dart
class class _ShellProxyCalls implements Shell {
  // Implements Shell by serializing messages and calling sendMessage on a _ShellProxyImpl.
}
```

```dart
class ShellProxy implements bindings.ProxyBase {
  final bindings.Proxy impl;
  // Mostly forwards calls to impl.
}
```

```dart
class ShellStub extends bindings.Stub {
  // Implements bindings.Stub.handleMessage for this specific interface.
  // Deserializes the message.
  // Invokes methods on an implementation of Shell.
}
```

**Comment by j9brown** [comment]
It sounds like it would be worth trying to find a more idiomatic or less
intrusive way of expressing these interfaces in Dart.  Possibly also a good
time to think about performance overhead from serialization or indirection.

**Comment by abarth** [comment]
Here's a proposal:

`s/bindings.Proxy/MojoMessageSender/`
`s/bindings.Stub/MojoMessageReceiver/`

`s/bindings.ProxyBase/bindings.Proxy/`
`abstract class Shell` <--- Keep
`class ShellStub` <--- Keep
`class ShellProxy` <--- Keep

`class _ShellProxyImpl` and `class _ShellProxyCalls` -> combine and rename to `_ShellMessageSender`

**Comment by abarth** [comment]
> When dealing with a Proxy you must access a field called "ptr" which I guess is short hand for "pointer", which has no meaning in Dart.

It's nice to have some way to get a clean object that has exactly the public interface.  Calling it `ptr` is kind of goofy, but I support the idea.

**Comment by zanderso** [comment]
We have to be careful with the `Calls` interface. The names of the methods there are generated from the public interface. We'd have to (further) mangle the generated method names to avoid conflicts if we combine or otherwise move those methods around. Totally on board with coming up with better names, especially for `ptr`, but we have to be very careful if we want to eliminate that indirection.

**Comment by jamesr** [comment]
Collisions are possible with any name generation scheme so it's useless to try to completely avoid them.  I think the only reasonable approach is to generate names that are as nice as possible and define a generic collision handling algorithm so only names that collide in practice have less-obvious names.

**Comment by abarth** [comment]
We can use mangled names in the MojoMessageSender / _ShellMessageSender and leave the unmangled names for the mojom-defined names.  For example, mojom could ban using names with leading or trailing underscores, which would let us use those names in the implementation.

**Comment by MitchRudominer** [comment]
I'm fine with specifying that Mojom names cannot begin with a leading underscore. Or we could say two underscores a la Python.

**Comment by zanderso** [comment]
Symbols with a leading underscore are library-private in Dart, so reserving names with a trailing underscore would give us more flexibility.

**Comment by floitschG** [comment]
It would be great if we could disallow symbols with leading underscores, as they would otherwise need to be mangled in some way. As Zach states: having access to names with _trailing_ underscores would gives some room for non-service names.

Here is my attempt of proposing a simpler hierarchy.

As a summary:
* `s/bindings.Proxy/ProxyMessageHandler/`  (although Adam's suggestion is good too. I just started from scratch and hadn't looked at his names).
* `s/bindings.Stub/StubMessageHandler/`
* The `ProxyMessageHandler`s below are fields instead of `extends`. The advantage of having them as fields is, that the handling methods aren't exposed to the user since they can stay private (they are just torn off and passed to the `ProxyMessageHandler`). I guess it would be possible to use inheritance, though.
* Create a `ProxyControl` class which has the methods that are needed for proxies, but are not part of the `Shell` interface.
* The `ShellProxy` then has a `proxyControl` field (the only non-`Shell` field) which gives access to the proxy functionality. If trailing underscores are reserved, then this should probably be `proxyControl_`.

``` dart
// ================= Proxy ========================

class ProxyMessageHandler extends core.MojoEventHandler {
  Function _handleResponse;
  Function _handleWrite;

  ProxyMessageHandler.unbound(void this._handleResponse(), void this._handleWrite())
      : super.unbound();

  // And similar constructors.

  // Replace calls to handleResponse with calls to _handleResponse, and
  // calls to handleWrite with calls to _handleWrite.
}

class ProxyControl {
  final String serviceName;

  Future close({bool immediate: false});
  Future get errorFuture;
  int get version;
  int requireVersion(...);
}

class Proxy {
  final ProxyControl proxyControl;

  Proxy(this.proxyControl);
}

// ================= Shell specific proxy

class Shell {
  // abstract shell;
}

class _ShellProxyControl implements ProxyControl {
  final ProxyMessageHandler _messageHandler;  // ProxyControl could also extend ProxyMessageHandler
  // Implements _handleResponse, and _handleWrite. which are given to the ProxyMessageHandler
}

class ShellProxy extends Proxy implements Shell {
  ShellProxy(...) : super(new _ShellProxyControl(..));

  // Forwards all Shell functions to proxyControl._messageHandler
}

// ===================== Stub ====================

/// Serves requests from a proxy by forwarding calls to the local backing `implementation`.
class Stub<T> {
  T implementation;
}

class StubMessageHandler extends core.MojoEventHandler {
  Function _handleMessage;

  StubMessageHandler.unbound(dynamic this._handleMessage(ServiceMessage message))
      : super.unbound();

  // Replace handleMessage calls with _handleMessage calls.
}

// =======Shell specific Stub.

class ShellStub extends Stub<Shell> {
  StubMessageHandler _handler = new StubMessageHandler(this._handleMessage);  // Doesn't work. I know.
  _handleMessage(...) { /* forwards to implementation */ }
}
```

I also wonder if we should make `responseOrError` be the default (only?) way to receive values.
It feels error-prone not to listen to the `errorFuture`...

**Comment by zanderso** [comment]
@floitschG Regarding:

> I also wonder if we should make responseOrError be the default (only?) way to receive values. It feels error-prone not to listen to the errorFuture...

 See: https://github.com/domokit/mojo/issues/673#issuecomment-186308189

**Comment by zanderso** [comment]
I've drawn up a change for Proxies that mostly[1] follows @floitschG's suggestions.

https://codereview.chromium.org/1964193002/

It has the advantage that `ptr` goes away, which seems to be the main annoyance. The interface calls are generated directly into the Proxy classes. This required moving the other calls that used to be in the Proxy classes into the ProxyControl classes, as suggested above, so it may be needed to write e.g. `proxy.ctrl.isBound` where before you'd have just written `proxy.isBound`. In porting over the examples and tests in the Mojo tree, this didn't seem too bad after punching holes for `close()` and `responseOrError()`.

If this sounds good, I'll proceed with this CL.

[1] Passing the `handleResponse` callback to the constructors of `ProxyMessageHandler` turned out to be too awkward, and so it made sense for the generated ProxyControl classes to inherit from `ProxyMessageHandler`. It was also easier for the `ProxyControl` base class to inherit most of its interface from `ProxyMessageHandler`.

**Comment by j9brown** [comment]
Sounds interesting.  I'll have to take a closer look.

Can we do something to improve the static typing in the generated
interfaces?  Last I checked there was a lot of dynamic and Object which
hurts API discoverability and likewise will limit the assistance an IDE can
provide during authoring.

**Comment by zanderso** [comment]
@j9brown Let's make a separate issue for that, and if you could put some specific examples there too, that would help a lot.

**Comment by abarth** [comment]
The specific problems that I've run into are that for mojom interface `Foo` that has a `bar` message that returns a result, the generated `Foo` class returns `dynamic` from `bar` and takes a mysterious optional factory parameter.  I think the core issue is that we're trying to share the `Foo` interface between the stub and proxy objects.  However, those objects (at least as currently implemented) actually have different interfaces.  To resolve this issue, we should either make them actually have the same interface (as in C++ or Java) or use different classes to represent their different interfaces.

I'd lean towards actually make them have the same interface, probably by dropping `Future`s from the proxy versions of the interface.  In practice, the generated interfaces aren't idiomatic enough to use directly in Dart code anyway because, in mojom, messages can return multiple results, which doesn't map well to `Future`s, which can resolve only to a single value.  The struct the generated bindings introduce to paper over this issue ends up not being something we want to expose to end-developers.  If we used callbacks instead of `Future`s, we could handle multiple results in the same way they're handled in other languages (as multiple parameters to the callback) and the stub and proxy side of the connection could actually have the same interface (again, as in other languages).

**Comment by zanderso** [comment]
@abarth To make sure I understand, you mean that:

```mojo
interface Foo {
  Bar() => (int32 a, int32 b, int32 c)
};
```

becomes the Dart interface:

```dart
abstract class Foo {
  void bar(void callback(int a, int b, int c));
}
```

On the proxy side, the callback is invoked when a reply is received. On the Stub side, the callback makes a Response object and sends it?

It's not very idiomatic, but if Mojo code is always wrapped up in a more idiomatic Dart library, I think it's okay. Passing callbacks like this also avoids an extra trip through the event loop, and so should be a tiny bit faster.

**Comment by zanderso** [comment]
I've split the callbacks/Futures stuff off into a separate issue. Let's use this issue to track refactoring Proxies and Stubs.

**Comment by j9brown** [comment]
Insofar as we're talking about Proxies and Stubs, I remember being confused
by the fact that some code handles proxy and stub objects interchangeably.
Both get passed around as Objects and then type tests are performed in the
core of the Mojo bindings for Dart.

This previously led to a bug where we couldn't forward interface requests
across calls because the bindings forgot to handle the case where a stub
(proxy?) may have appeared in one spot.  All in all, I think adopting a
more rigorous static type discipline will be beneficial.  That is to say,
where necessary, clearly distinguishing between objects which represent
interfaces (proxies) vs. interface requests (stubs), as is done in the C++
layer, and otherwise use handles for more generic APIs.

I also noticed that the Mojo message pipe endpoints are being wired up to
the event loop eagerly upon deserialization.  In the general case, we don't
know whether the endpoint will actually be used locally by the program or
just forwarded somewhere else.  It would be good to defer this wiring until
something actually gets bound to the pipe endpoint.

**Comment by zanderso** [comment]
> Both get passed around as Objects and then type tests are performed in the
> core of the Mojo bindings for Dart.

My CL should help with this since the class hierarchy is simpler. I'll investigate after it lands.

> I also noticed that the Mojo message pipe endpoints are being wired up to
> the event loop eagerly upon deserialization.

The last non-benign instance of this was removed by https://github.com/domokit/mojo/commit/06721d95d3b55c26b50e8a0c1deb40c94bfced38.

**Comment by floitschG** [comment]
@abarth I kind of like the Futures. They make things simpler.

I would simply introduce the structs (that combine multiple return values) only for members that return multiple values. With a good name I think that solution would actually be nice.

I went through all (~150) mojom files in the mojo github repository, and these are the ones that use multiple return values (see below). Some of them could be "solved" if we had a convention of how to return errors. Currently they are sometimes the first return value and sometimes the last return value. If there was an agreed upon naming scheme (and position), then Dart could maybe complete the future with the error value (at least for the proxy case).

https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/shell/test/pingable.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/services/js/test/network_test_service.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/network/interfaces/host_resolver.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/native_support/interfaces/process.mojom
https://github.com/domokit/mojo/blob/db501eaa0c950caf6ab4bfae49ca5d1e9dfc0d60/mojo/services/media/core/interfaces/timeline_controller.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/network/interfaces/http_connection.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/network/interfaces/tcp_server_socket.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/terminal/interfaces/terminal.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/authentication/interfaces/authentication.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/files/interfaces/directory.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/location/interfaces/geocoder.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/media/control/interfaces/media_player.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/network/interfaces/udp_socket.mojom
https://github.com/domokit/mojo/blob/621ce2c2e4a22b5315deae70f214714717033f4f/mojo/services/media/core/interfaces/media_demux.mojom
https://github.com/domokit/mojo/blob/a0d454d213dbea24df0ade123c1b2837d53c8b8a/mojo/services/media/control/interfaces/media_sink.mojom
https://github.com/domokit/mojo/blob/fc4917ebe69b1251ccad1dbe21d0ea0a9c15f168/mojo/services/media/core/interfaces/seeking_reader.mojom
https://github.com/domokit/mojo/blob/a0d454d213dbea24df0ade123c1b2837d53c8b8a/mojo/services/media/control/interfaces/media_source.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/network/interfaces/network_service.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/url_response_disk_cache/interfaces/url_response_disk_cache.mojom
https://github.com/domokit/mojo/blob/df314089817655626186198e87189f788c86ee41/mojo/services/files/interfaces/file.mojom

**Comment by abarth** [comment]
@floitschG See https://github.com/domokit/mojo/issues/777#issuecomment-218637857 for a detailed discussion of the various options.

**Comment by zanderso** [comment]
With https://github.com/domokit/mojo/commit/02596a0415f2e785ce2475049d02c368e49a33ad, I've added generated `Interface` and `InterfaceRequest` types that will hopefully reduce the need to use Proxies and Stubs directly, and make the bindings less confusing to use.

Suppose you have a Mojo interface:

```mojo
interface Foo {
  DoFoo() => (int32 foo);
  RequestFoo(Foo& foo);
  SetFoo(Foo foo);
};
```

We'll generate the Dart interface:
```dart
abstract class Foo {
  dynamic doFoo([Function responseFactory]);
  void requestFoo(FooInterfaceRequest foo);
  void setFoo(FooInterface foo);
}
```

Which you can use as follows:

```dart
// Request the Foo service from the shell.
var foo = Foo.connectToService(mojoApplication, "mojo:foo");
var foos = await foo.doFoo();

// Request a Foo service by sending a Mojo interface request to
// Foo.RequestFoo().
var requestedFoo = new FooInterfaceRequest();
foo.requestFoo(requestedFoo);
var requestedFoos = await requestedFoo.doFoo();

// Send a Foo interface as a Mojo interface parameter.
var myFoo = new FooImpl();
var fooInterface = new FooInterface(myFoo);
requestedFoo.setFoo(fooInterface);
```

As for implementing `Foo`:
```dart
class FooImpl implements Foo {
  ...
  void requestFoo(FooInterfaceRequest foo) {
    // Provide an implementation:
    foo.impl = new FooImpl();
    // Or forward the request on:
   someOtherFoo.requestFoo(foo);
  }

  void setFoo(FooInterface foo) {
    // Use foo:
    foo.doFoo();
    // Or forward the interface on:
    someOtherFoo.setFoo(foo);
  }
}
```

The generated Interface and InterfaceRequest interfaces are implemented in terms of the generated Proxy and Stub classes, so existing code will continue to work.

I think the bindings are now cleaned up enough that we can close this issue. If there are new specific issues to address, please open a new issue.

**Comment by floitschG** [comment]
Would it be possible to type the responseFactory?
`dynamic doFoo([void responseFactory(int foo)]);`

**Comment by zanderso** [comment]
The consensus seemed to be to move forward with https://github.com/domokit/mojo/issues/777, eliminating the responseFactory entirely. I think I should probably start work on that next.

**Comment by floitschG** [comment]
Sgtm. thanks.

**Comment by j9brown** [comment]
Awesome, thanks!
